test/integration/specs/kbaseCellToolbarMenu_test.js

```diff
@@ -0,0 +1,327 @@
+
+        // Find and click the move-down button
+        const downButton = await cell.$('[data-test="cell-move-down"]');
+        await clickWhenReady(downButton);
+        await waitForCellWithTitle(narrativeContainer, testCase.cellIndex + 1, testCase.title);
+    });
+
+    it('moves a minimized unselected cell down', async () => {
+        const testCase = testCases.TEST_CASE2;
+        await login();
+        const narrativeContainer = await openNarrative(testCase.narrativeId);
+
+        const cell = await waitForCellWithTitle(narrativeContainer, testCase.cellIndex, testCase.title);
+
+        // Find and click the move-down button
+        const downButton = await cell.$('[data-test="cell-move-down"]');
+        await clickWhenReady(downButton);
+        await waitForCellWithTitle(narrativeContainer, testCase.cellIndex + 1, testCase.title);
+    });
+
+    it('moves a minimized unselected cell up', async () => {
+        const testCase = testCases.TEST_CASE3;
+        await login();
+        const narrativeContainer = await openNarrative(testCase.narrativeId);
+
+        const cell = await waitForCellWithTitle(narrativeContainer, testCase.cellIndex, testCase.title);
+
+        // Find and click the move-ups button
+        const upButton = await cell.$('[data-test="cell-move-up"]');
+        await clickWhenReady(upButton);
+        await waitForCellWithTitle(narrativeContainer, testCase.cellIndex - 1, testCase.title);
+    });
+
+    // select cell
+    it('selects a minimized cell', async () => {
+        const testCase = testCases.TEST_CASE4;
+        await login();
+        const narrativeContainer = await openNarrative(testCase.narrativeId);
+        await selectCell(narrativeContainer, testCase.cellIndex, testCase.title);
+    });
+
+    // select cell and move down
+    it('selects a minimized cell and moves it down', async () => {
+        const testCase = testCases.TEST_CASE4;
+        await login();
+        const narrativeContainer = await openNarrative(testCase.narrativeId);
+        const cell = await selectCell(narrativeContainer, testCase.cellIndex, testCase.title);
+        const downButton = await cell.$('[data-test="cell-move-down"]');
+        await clickWhenReady(downButton);
+        await waitForCellWithTitle(narrativeContainer, testCase.cellIndex + 1, testCase.title);
+    });
+
+    // select cell and move up
+    it('selects a minimized cell and moves it up', async () => {
+        const testCase = testCases.TEST_CASE4;
+        await login();
+        const narrativeContainer = await openNarrative(testCase.narrativeId);
+        const cell = await selectCell(narrativeContainer, testCase.cellIndex, testCase.title);
+        const upButton = await cell.$('[data-test="cell-move-up"]');
+        await clickWhenReady(upButton);
+        await waitForCellWithTitle(narrativeContainer, testCase.cellIndex - 1, testCase.title);
+    });
+
+    // Everything above, but cells are expanded.
```

Comment:
If you're repeating a set of tests with a single variable changed, I would rewrite my testing functions to parameterise the variable; it makes the code easier to maintain and there's less chance of a copy-paste error. ISTM you could write a more generic testing function that takes in four params--expanded/collapsed state, selection state, and movement direction, plus test case data--and use that to test all combinations of parameters.

Each test also starts with `await login()`, so that can go into the `beforeEach()` function.

Reply:
I did quite a round of test refactoring to be more DRY, and fixed some false test positives.
Basically, I was getting stretched by the amount of test time to actual bug fixing time, and just stopped when the tests reported success (and a round or two of refactoring them.) That is partly because the integration testing is so new, and we are still building up support and getting experience.